### ISSUE TITLE
BUG: PyObject_IsTrue error checking

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#24177]
While reviewing #24175 I noticed that we're not consistently checking if `PyObject_IsTrue` returns -1:

```
nathan at rous in ~/Documents/numpy on flatiter-copyswap!
± rg PyObject_IsTrue
numpy/core/src/umath/loops.c.src
2296:            int ret = PyObject_IsTrue(ret_obj);

numpy/core/src/umath/funcs.inc.src
93:        int retcode = PyObject_IsTrue(i1);
119:        int retcode = PyObject_IsTrue(i1);

numpy/core/src/common/npy_ctypes.h
34:    ret = PyObject_IsTrue(ret_obj);

numpy/core/src/multiarray/iterators.c
618:        if (PyObject_IsTrue(ind)) {
918:        if (PyObject_IsTrue(ind)) {

numpy/core/src/multiarray/alloc.c
71:    int enabled = PyObject_IsTrue(enabled_obj);

numpy/core/src/multiarray/methods.c
2759:        int istrue = PyObject_IsTrue(uic);
2778:        int istrue = PyObject_IsTrue(write_flag);

numpy/core/src/multiarray/arraytypes.c.src
325: * #func2 = PyObject_IsTrue, MyPyLong_AsLong*6, MyPyLong_AsUnsignedLong*2,
2847:        return (npy_bool) PyObject_IsTrue(*ip);
2855:        return (npy_bool) PyObject_IsTrue(obj);

numpy/core/src/multiarray/mapping.c
491:                if (PyObject_IsTrue((PyObject *)arr)) {

numpy/core/src/multiarray/sequence.c
49:    ret = PyObject_IsTrue(any);

numpy/core/src/multiarray/ctors.c
2272:        if (PyObject_IsTrue(PyTuple_GET_ITEM(attr,1))) {

numpy/core/src/multiarray/flagsobject.c
282:                              (PyObject_IsTrue(obj) ? Py_True : Py_False));
307:                              (PyObject_IsTrue(obj) ? Py_True : Py_False),
333:                              (PyObject_IsTrue(obj) ? Py_True : Py_False),
356:    ret = PyObject_IsTrue(obj);

numpy/f2py/rules.py
921:        #varname# = (#ctype#)PyObject_IsTrue(#varname#_capi);

numpy/core/src/multiarray/_multiarray_tests.c.src
936:    writeable = PyObject_IsTrue(arg);
1048:    replace = PyObject_IsTrue(replace_obj);

numpy/core/src/multiarray/conversion_utils.c
430:    if (PyObject_IsTrue(object)) {
```

All of the occurances where there's no checking of the return value probably need error checking, especially if the object is user-supplied, although since the truthyness of ndarrays is poorly defined this also bites for numpy types too.

**Comment by seberg** [comment]
Seems like an easy thing to work on for new-comers gh-24178 shows a bit of the how, although some of these would probably be nicer fixed by integrating this into early argument parsing. (and some are correct)

Adding tests where easy would be nice.  (some may well not be possible to trigger in practice)

**Comment by abdulasiraj** [comment]
Hi @seberg,
can you please assign this to me?
Thanks!

**Comment by seberg** [comment]
We don't assign issues.  Please just open a PR that references this issues so others will see that you started working on it.

**Comment by abdulasiraj** [comment]
Sure. I'll refer PR. Thanks!

**Comment by ngoldbaum** [comment]
This was fixed by #24237 but the issue never got closed.
